src/mumble/GlobalShortcut_win.cpp

```diff
@@ -130,576 +52,356 @@ GlobalShortcutWin::~GlobalShortcutWin() {
-			UnhookWindowsHookEx(hhKeyboard);
-		}
+	if (!RegisterRawInputDevices(rid, nRid, sizeof(RAWINPUTDEVICE))) {
+		qWarning("GlobalShortcutWindows: RegisterRawInputDevices() failed with error %u", GetLastError());
 	}
 
-	foreach (InputDevice *id, qhInputDevices) {
-		if (id->pDID) {
-			id->pDID->Unacquire();
-			id->pDID->Release();
-		}
-		delete id;
-	}
-	pDI->Release();
-}
+	QTimer timer;
+	connect(&timer, &QTimer::timeout, this, &GlobalShortcutWin::timeTicked);
+	timer.start(20);
 
-bool GlobalShortcutWin::event(QEvent *event) {
-	QEvent::Type type = event->type();
-	if (type == INJECTKEYBOARDMESSAGE_QEVENT) {
-		InjectKeyboardMessageEvent *ikme = static_cast< InjectKeyboardMessageEvent * >(event);
-		bool suppress = handleKeyboardMessage(ikme->m_scancode, ikme->m_vkcode, ikme->m_extended, ikme->m_down);
-		ikme->m_suppressionPromise.set_value(suppress);
-		return true;
-	} else if (type == INJECTMOUSEMESSAGE_QEVENT) {
-		InjectMouseMessageEvent *imme = static_cast< InjectMouseMessageEvent * >(event);
-		bool suppress                 = handleMouseMessage(imme->m_btn, imme->m_down);
-		imme->m_suppressionPromise.set_value(suppress);
-		return true;
-	}
-	return GlobalShortcutEngine::event(event);
+	exec();
 }
 
-bool GlobalShortcutWin::injectKeyboardMessage(MSG *msg) {
-	if (!bHook) {
-		return false;
+void GlobalShortcutWin::injectRawInputMessage(HRAWINPUT handle) {
+	UINT size;
+	if (GetRawInputData(handle, RID_INPUT, nullptr, &size, sizeof(RAWINPUTHEADER)) != 0) {
+		return;
 	}
 
-	// Only allow keyboard messages.
-	switch (msg->message) {
-		case WM_KEYDOWN:
-		case WM_KEYUP:
-		case WM_SYSKEYDOWN:
-		case WM_SYSKEYUP:
-			break;
-		default:
-			return false;
+	auto buffer = std::make_unique< uint8_t[] >(size);
+	if (GetRawInputData(handle, RID_INPUT, buffer.get(), &size, sizeof(RAWINPUTHEADER)) <= 0) {
+		return;
 	}
 
-	DWORD scancode = (msg->lParam >> 16) & 0xff;
-	DWORD vkcode   = msg->wParam;
-	bool extended  = !!(msg->lParam & 0x01000000);
-	bool up        = !!(msg->lParam & 0x80000000);
-
-	InjectKeyboardMessageEvent *ikme = new InjectKeyboardMessageEvent(scancode, vkcode, extended, !up);
-	boost::future< bool > suppress   = ikme->shouldSuppress();
-	qApp->postEvent(this, ikme);
-	return suppress.get();
-}
-
-bool GlobalShortcutWin::injectMouseMessage(MSG *msg) {
-	if (!bHook) {
-		return false;
-	}
+	auto input = reinterpret_cast< const PRAWINPUT >(buffer.get());
+	switch (input->header.dwType) {
+		case RIM_TYPEMOUSE: {
+			const auto &mouse = input->data.mouse;
+			emit mouseMessage(mouse.usButtonFlags, mouse.usButtonData);
+			break;
+		}
+		case RIM_TYPEKEYBOARD: {
+			const auto &keyboard = input->data.keyboard;
+			if (keyboard.MakeCode == KEYBOARD_OVERRUN_MAKE_CODE) {
+				// Invalid or unrecognizable combination of keys is pressed or
+				// the number of keys pressed exceeds the limit for this keyboard.
+				return;
+			}
 
-	bool down        = false;
-	unsigned int btn = 0;
+			if (keyboard.VKey == 255) {
```

Comment:
Is there no constant defined inside the RawInput API that we can use for this comparison instead of the literal `255`?

Reply:
Unfortunately not, the defines stop at `0xFC`: https://docs.microsoft.com/en-us/windows/win32/inputdev/virtual-key-codes